### DISCUSSION TITLE
Use local files instead of Uri for map icon image

### DISCUSSION THREAD

**Discussion by davefxy** [VladislavAntonyuk/MauiSamples#150] [Q&A] [answered]
To use local files instead of Uri remote images for map icon images on an Android , I made the following changes to the MauiMaps sample. I made the following changes to the CustomMapHandler.cs AddPins method:
```csharp
               if (pin is MapLocation cp)
                {
                    Bitmap bitmap = null;
                    if (cp.ImageSource is FileImageSource)
                    {
                        var imageFileSourceHandler = new FileImageSourceHandler();
                        bitmap = await imageFileSourceHandler.LoadImageAsync(cp.ImageSource, Application.Context);
                    }
                    else
                    {
                        var imageUriSourceHandler = new ImageLoaderSourceHandler();
                        bitmap = await imageUriSourceHandler.LoadImageAsync(cp.ImageSource, Application.Context);
                    }
                    markerOption?.SetIcon(bitmap is null
                                              ? BitmapDescriptorFactory.DefaultMarker()
                                              : BitmapDescriptorFactory.FromBitmap(bitmap));
                }
```
But I dont know how to make these changes for iOS or MacCatalyst.

**Comment by VladislavAntonyuk** [comment, marked as answer]
you don't need to change Handler and use FileImageSourceHandler. It should automatically use correct handler to load image from the ImageSource depends on its type. You need to provide a valid ImageSource to make it work

**Comment by davefxy** [comment]
I had to make the changes I described for the Android because I was getting null for the bitmap using ImageLoaderSourceHandler() to try to load a local file for the pin image. The changes allow it to work. So the Android must not be working as you expect.
The image files (PNG) were downloaded and stored locally because the website the images are downloaded from is an HTTPS website that requires a User-Agent header field. So I cannot use the image's Url as an imagesource in the pin definition..
Another interesting question is how to resize the downloaded image and what should the new size be. In Xamarin we used to be able to query the platform what the scale factor is. I could then have a platform specific method for resizing images before I stored it locally.

**Reply by VladislavAntonyuk** [reply]
Most likely you put invalid path to the image. I will update sample for you with local image pin as well

**Reply by VladislavAntonyuk** [reply]
Updated

**Comment by davefxy** [comment]
I see that you are passing a memory stream. I was passing a file name
. Your method:
                ImageSource = ImageSource.FromStream(() => new MemoryStream(File.ReadAllBytes(ImagePath))),
   my method:
                ImageSource = ImageSource.FromFile(ImagePath),
With my method, I had to add a file type check in the platform handler.

**Reply by VladislavAntonyuk** [reply]
this code is cross-platformed, you don't need to add additional logic to Handler. You only need to specify a valid path for the image. it can be done in MainPage.xaml.cs. You can use `Environment.GetFolderPath(...)`, use the MAUI FileSystem class, or create your own FileSystemService that will return you the path to the image. it is out of the scope of the Maps sample
